### PR TITLE
test get/scan without family

### PR DESCRIPTION
**Pull request by foronedream**
table group search 
## Summary

table group search 

## Solution Description

table group search

### CODE REVIEW

File: src/test/java/com/alipay/oceanbase/hbase/HTableTestBase.java [hunk 1]
```diff
@@ -46,6 +47,49 @@ public abstract class HTableTestBase {
...
+        String column1 = "putColumn3";
+        String value = "value333444";
+        long timestamp = 170558 * 10000000 + 5262338;
+        // put data
+        Put put = new Put(toBytes(key));
+        put.add("family1".getBytes(), column1.getBytes(), timestamp, toBytes(value + "1"));
+        hTable.put(put);
+        
+        Get get = new Get(toBytes(key));
+        // get.addColumn("family1".getBytes(), toBytes(column1));
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/test/java/com/alipay/oceanbase/hbase/HTableTestBase.java]: done

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 2]
```diff
@@ -361,33 +361,54 @@ public void batch(List<? extends Row> actions, Object[] results) {
...
+                index = i;
+                break;
+            }
+        }
+        int len = tableName.length - index - 1;
+        // Family and column, return array size 2
+        final byte[] result = new byte[len];
+        System.arraycopy(tableName, index + 1 /* Skip delimiter */, result, 0, len);
+        return result;
+    }
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java]: done

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 3]
```diff
@@ -361,33 +361,54 @@ public void batch(List<? extends Row> actions, Object[] results) {
...
+        }
 
         ServerCallable<Result> serverCallable = new ServerCallable<Result>(configuration,
             obTableClient, tableNameString, get.getRow(), get.getRow(), operationTimeout) {
             public Result call() throws IOException {
                 List<KeyValue> keyValueList = new ArrayList<KeyValue>();
-                for (Map.Entry<byte[], NavigableSet<byte[]>> entry : get.getFamilyMap().entrySet()) {
-
-                    byte[] family = entry.getKey();
+                if (obTableClient.isTableGroup()) {
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java]: done

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 4]
```diff
@@ -403,6 +424,39 @@ public Result call() throws IOException {
                         throw new IOException("query table:" + tableNameString + " family "
                                               + Bytes.toString(family) + " error.", e);
                     }
+                } else {
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java]: done

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 5]
```diff
@@ -365,44 +366,114 @@ public Object[] batch(List<? extends Row> actions) {
...
+        int qualifierLen = total_length - familyLen - 1;
+        byte[] newQualifier = new byte[qualifierLen];
+        if (qualifierLen > 0) {
+            System.arraycopy(qualifier, familyLen + 1 /* Skip delimiter */, newQualifier, 0, qualifierLen);
+        } else {
+            throw new RuntimeException("can not get qualifier name");
+        }
+        res[1] = newQualifier;
+        return res;
+    }
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/hbase/OHTable.java:397]:
```suggestion
public byte[][] extractFamilyFromQualifier(byte[] qualifier) {
    int familyLen = -1;
    for (int i = 0; i < qualifier.length; i++) {
        if (qualifier[i] == '\0') {
            familyLen = i;
            break;
        }
    }
    if (familyLen == -1) {
        throw new RuntimeException("Cannot get family name");
    }

    byte[] family = Arrays.copyOfRange(qualifier, 0, familyLen);
    int qualifierLen = qualifier.length - familyLen - 1;
    if (qualifierLen <= 0) {
        throw new RuntimeException("Cannot get qualifier name");
    }

    byte[] newQualifier = Arrays.copyOfRange(qualifier, familyLen + 1, qualifier.length);
    return new byte[][] { family, newQualifier };
}
```

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 6]
```diff
@@ -365,44 +366,114 @@ public Object[] batch(List<? extends Row> actions) {
...
 
-        checkFamilyViolation(get.getFamilyMap().keySet());
+    public void getKeyValueFromResult(ObTableClientQueryStreamResult clientQueryStreamResult, 
+                                       List<KeyValue> keyValueList, boolean isTableGroup, byte[] family) throws Exception {
+        byte[][] familyAndQualifier = new byte[2][];
+        for (List<ObObj> row : clientQueryStreamResult.getCacheRows()) {
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/hbase/OHTable.java:402]: Try using `getRow()` and `next()`

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java [hunk 7]
```diff
@@ -365,44 +366,114 @@ public Object[] batch(List<? extends Row> actions) {
...
+            System.arraycopy(qualifier, familyLen + 1 /* Skip delimiter */, newQualifier, 0, qualifierLen);
+        } else {
+            throw new RuntimeException("can not get qualifier name");
+        }
+        res[1] = newQualifier;
+        return res;
+    }
 
-        checkFamilyViolation(get.getFamilyMap().keySet());
+    public void getKeyValueFromResult(ObTableClientQueryStreamResult clientQueryStreamResult,
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/hbase/OHTable.java:399]: What about ObTableClientQueryAsyncStreamResult

File: src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java [hunk 8]
```diff
@@ -41,17 +41,50 @@ public class ClientStreamScanner extends AbstractClientScanner {
...
     public ClientStreamScanner(ObTableClientQueryStreamResult streamResult, String tableName,
-                               byte[] family) {
+                               byte[] family, boolean isTableGroup) {
         this.streamResult = streamResult;
         this.tableName = tableName;
         this.family = family;
+        this.isTableGroup = isTableGroup;
+    }
+
+    public byte[][] extractFamilyFromQualifier(byte[] qualifier) throws Exception {
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java:60]: Try using one method

File: src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java [hunk 9]
```diff
@@ -73,8 +106,19 @@ public Result next() throws IOException {
                 return null;
             }
 
+
+            byte[][] familyAndQualifier = new byte[2][];
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java:110]: I guess `byte[][] familyAndQualifier;` is ok

File: src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java [hunk 10]
```diff
@@ -73,8 +106,19 @@ public Result next() throws IOException {
...
+            if (this.isTableGroup) {
+                // split family and qualifier
+                familyAndQualifier = extractFamilyFromQualifier((byte[]) startRow.get(1).getValue());
+                this.family = familyAndQualifier[0];
+            } else {
+                familyAndQualifier[1] = (byte[]) startRow.get(1).getValue();
+            }
+
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java:118]: Remove useless blank line

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java (hunk 7 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java:372]: A single ObTableClientQueryAsyncStreamResult handles the same, use next() and getRow().

File: src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java (hunk 8 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java]: add utils file

File: src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java (hunk 9 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java:81]: no， Two-dimensional array needs to specify the dimension

File: src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java (hunk 10 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/result/ClientStreamScanner.java:89]: done

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java (hunk 6 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java]: done

File: src/main/java/com/alipay/oceanbase/hbase/OHTable.java (hunk 5 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/OHTable.java]: done

File: src/main/java/com/alipay/oceanbase/hbase/util/OHBaseFuncUtils.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
+package com.alipay.oceanbase.hbase.util;
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/hbase/util/OHBaseFuncUtils.java:1]: license

Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/hbase/util/OHBaseFuncUtils.java:18]: done

**Comment by IHEII** [comment]
LGTM

**Pull request merged**